### PR TITLE
Coding Exercise Part 1

### PR DESCRIPTION
**Pull request by jamiesanerd**
(video in email)

## Summary by CodeRabbit

- **New Features**
  - Introduced a `PurchaseOrdersService` to fetch all purchase orders sorted by expected delivery date.
  - Added a `ShopifyPurchaseOrdersTable` component for displaying and managing purchase orders.

- **Enhancements**
  - Updated client app layout with new styling and structure.
  - Integrated `ShopifyPurchaseOrdersTable` in the purchase orders page.

- **Dependencies**
  - Updated package dependencies including `@shopify/polaris` and `browserslist`.

- **Configuration**
  - Updated Node.js version to `v20.10.0` in `.nvmrc`.

### CODE REVIEW

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+  const resourceName = {
+    singular: 'order',
+    plural: 'orders',
+  };
+
+  const [sortDirection, setSortDirection] = useState<IndexTableSortDirection>();
+  const [sortedColumnIndex, setSortedColumnIndex] = useState<number>();
+
+  const handleSort = (
```
Comment (johngully) [inline apps/client/components/ShopifyPurchaseOrdersTable.tsx:74]: Did you consider handling sorting on the server using Prisma?  How would this work with larger datasets?

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                key={id}
+                selected={selected}
+                position={index}
+              >
+                <IndexTable.Cell>
+                  <Text as="span">{id}</Text>
+                </IndexTable.Cell>
+                <IndexTable.Cell>{vendor_name}</IndexTable.Cell>
+                <IndexTable.Cell>
+                  {' '}
```
Comment (johngully) [inline apps/client/components/ShopifyPurchaseOrdersTable.tsx:141]: Why is an extra whitespace character being used here?

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                  </Text>
+                </IndexTable.Cell>
+
+                <IndexTable.Cell>
+                  <Text as="span" alignment="end" numeric>
+                    {quantity}
+                  </Text>
+                </IndexTable.Cell>
+                <IndexTable.Cell>
+                  <Text as="span" alignment="end" numeric>
```
Comment (johngully) [inline apps/client/components/ShopifyPurchaseOrdersTable.tsx:159]: Did you consider adding components for each datatype that would handle `date`, `numeric`, `string` data and apply the necessary formatting consistently

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx
Comment (johngully) [inline apps/client/components/ShopifyPurchaseOrdersTable.tsx:1]: This component is getting pretty big, what would you do to make this more readable, composable, etc

**Review by johngully — COMMENTED** [review]
Any chance we can get a video walkthrough of your work here?  What was easy/hard, how did you make the design decisions you made? What are your thoughts on how this can be reused across the application?

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx (hunk 3 above)
Comment (jamiesanerd) [reply apps/client/components/ShopifyPurchaseOrdersTable.tsx:159]: For sure with future passes I would decompose this into row/cell components once I'd finialized on the functionality. Generally I'll let a first pass get a bit verbose so I can see the whole picture before deciding to split off a subset of components.

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx (hunk 1 above)
Comment (jamiesanerd) [reply apps/client/components/ShopifyPurchaseOrdersTable.tsx:74]: If i was going to be working with more data I would likely move this serverside with a couple enahncements:
- pagination
- query params
- sort params
Then I could pass my current page, any sorting/query options i've done and update my state based on the response from the new query. tradeoffs here would be drastically increasing server calls at the benefit of improving page performance

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx
Comment (jamiesanerd) [reply apps/client/components/ShopifyPurchaseOrdersTable.tsx:1]: as above, moving row/cell out into their own generic elements would be my next step

File: apps/client/components/ShopifyPurchaseOrdersTable.tsx (hunk 2 above)
Comment (jamiesanerd) [reply apps/client/components/ShopifyPurchaseOrdersTable.tsx:141]: I could honestly remove this, this was a leftover from working out how polaris worked and their Text element didn't play well with potentially null values, could easily address this in a cleaner way in a DateCell component

**Pull request closed without merging**